File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/proxy/RtpTransceiverProxy.kt
package com.instrumentisto.medea_flutter_webrtc.proxy

import com.instrumentisto.medea_flutter_webrtc.model.RtpTransceiverDirection
import org.webrtc.RtpTransceiver

/** Wrapper around an [RtpTransceiver]. */
class RtpTransceiverProxy(obj: RtpTransceiver) : Proxy<RtpTransceiver>(obj) {
  /** [RtpSenderProxy] of this [RtpTransceiverProxy]. */
  private lateinit var sender: RtpSenderProxy

  /** [RtpReceiverProxy] of this [RtpTransceiverProxy]. */
  private lateinit var receiver: RtpReceiverProxy

  /** TODO */
  private var disposed: Boolean = false

  /** TODO */
  private var mid: String? = null

  init {
    syncSender()
    syncReceiver()
    addOnSyncListener {
      syncSender()
      syncReceiver()
    }
  }

  /** TODO */
  fun setDisposed() {
    disposed = true
    receiver.setDisposed()
    sender.setDisposed()
  }

  /** @return [RtpSenderProxy] of this [RtpTransceiverProxy]. */
  fun getSender(): RtpSenderProxy {
    return sender
  }

  /** @return [RtpReceiverProxy] of this [RtpTransceiverProxy]. */
  fun getReceiver(): RtpReceiverProxy {
    return receiver
  }

  /** Sets [RtpTransceiverDirection] of the underlying [RtpTransceiver]. */
  fun setDirection(direction: RtpTransceiverDirection) {
    if (!disposed) {
      obj.direction = direction.intoWebRtc()
    }
  }

  /** Sets receive of the underlying [RtpTransceiver]. */
  fun setRecv(recv: Boolean) {
    if (!disposed) {
      var currentDirection = RtpTransceiverDirection.fromWebRtc(obj)
      var newDirection =
          if (recv) {
            when (currentDirection) {
              RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.RECV_ONLY
              RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.RECV_ONLY
              RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_RECV
              RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_RECV
              else -> {
                RtpTransceiverDirection.STOPPED
              }
            }
          } else {
            when (currentDirection) {
              RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.INACTIVE
              RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.INACTIVE
              RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_ONLY
              RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_ONLY
              else -> {
                RtpTransceiverDirection.STOPPED
              }
            }
          }
      if (newDirection != RtpTransceiverDirection.STOPPED) {
        setDirection(newDirection)
      }
    }
  }

  /** Sets send of the underlying [RtpTransceiver]. */
  fun setSend(send: Boolean) {
    if (!disposed) {
      var currentDirection = RtpTransceiverDirection.fromWebRtc(obj)
      var newDirection =
          if (send) {
            when (currentDirection) {
              RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.SEND_ONLY
              RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_ONLY
              RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_RECV
              RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.SEND_RECV
              else -> {
                RtpTransceiverDirection.STOPPED
              }
            }
          } else {
            when (currentDirection) {
              RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.INACTIVE
              RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.INACTIVE
              RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.RECV_ONLY
              RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.RECV_ONLY
              else -> {
                RtpTransceiverDirection.STOPPED
              }
            }
          }
      if (newDirection != RtpTransceiverDirection.STOPPED) {
        setDirection(newDirection)
      }
    }
  }

  /** @return mID of the underlying [RtpTransceiver]. */
  fun getMid(): String? {
    if (!disposed) {
      mid = obj.mid
    }
    return mid
  }

  /** @return Preferred [RtpTransceiverDirection] of the underlying [RtpTransceiver]. */
  fun getDirection(): RtpTransceiverDirection {
    if (!disposed) {
      return RtpTransceiverDirection.fromWebRtc(obj)
    }
    return RtpTransceiverDirection.STOPPED
  }

  /** Stops the underlying [RtpTransceiver]. */
  fun stop() {
    receiver.notifyRemoved()
    if (!disposed) {
      obj.stop()
    }
  }

  /**
   * Synchronizes the [RtpSenderProxy] of this [RtpTransceiverProxy] with the underlying
   * [RtpTransceiver].
   */
  private fun syncSender() {
    val newSender = obj.sender
    if (this::sender.isInitialized) {
      sender.replace(newSender)
    } else {
      sender = RtpSenderProxy(newSender)
    }
  }

  /**
   * Synchronizes the [RtpReceiverProxy] of this [RtpTransceiverProxy] with the underlying
   * [RtpTransceiver].
   */
  private fun syncReceiver() {
    val newReceiver = obj.receiver
    if (this::receiver.isInitialized) {
      receiver.replace(newReceiver)
    } else {
      receiver = RtpReceiverProxy(newReceiver)
    }
  }
}

File: ios/Classes/proxy/RtpTransceiverProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpTransceiver`.
class RtpTransceiverProxy {
  /// Actual underlying `RTCRtpTransceiver`.
  private var transceiver: RTCRtpTransceiver

  /// `RtpSenderProxy` of this `RtpTransceiverProxy`.
  private var sender: RtpSenderProxy

  /// `RtpReceiverProxy` of this `RtpTransceiverProxy`.
  private var receiver: RtpReceiverProxy

  /// Creates a new `RtpTransceiverProxy` for the provided `RTCRtpTransceiver`
  init(transceiver: RTCRtpTransceiver) {
    self.sender = RtpSenderProxy(sender: transceiver.sender)
    self.receiver = RtpReceiverProxy(receiver: transceiver.receiver)
    self.transceiver = transceiver
  }

  /// - Returns: `RtpSenderProxy` of this `RtpTransceiverProxy`.
  func getSender() -> RtpSenderProxy {
    self.sender
  }

  /// - Returns: `RtpReceiverProxy` of this `RtpTransceiverProxy`.
  func getReceiver() -> RtpReceiverProxy {
    self.receiver
  }

  /// Sets `TransceiverDirection` of the underlying `RtpTransceiver`.
  func setDirection(direction: TransceiverDirection) {
    self.transceiver.setDirection(direction.intoWebRtc(), error: nil)
  }

  /**
    Sets recv state of this `RtpTransceiver`.

    - Parameters:
      - recv: Is receiver direction is enabled.
  */
  func setRecv(recv: Bool) {
    let direction = self.getDirection()
    var newDirection = RTCRtpTransceiverDirection.stopped
    if recv {
      switch direction {
      case .inactive:
        newDirection = RTCRtpTransceiverDirection.recvOnly
      case .recvOnly:
        newDirection = RTCRtpTransceiverDirection.recvOnly
      case .sendRecv:
        newDirection = RTCRtpTransceiverDirection.sendRecv
      case .sendOnly:
        newDirection = RTCRtpTransceiverDirection.sendRecv
      case .stopped:
        newDirection = RTCRtpTransceiverDirection.stopped
      }
    } else {
      switch direction {
      case .inactive:
        newDirection = RTCRtpTransceiverDirection.inactive
      case .recvOnly:
        newDirection = RTCRtpTransceiverDirection.inactive
      case .sendRecv:
        newDirection = RTCRtpTransceiverDirection.sendOnly
      case .sendOnly:
        newDirection = RTCRtpTransceiverDirection.sendOnly
      case .stopped:
        newDirection = RTCRtpTransceiverDirection.stopped
      }
    }

    if newDirection != RTCRtpTransceiverDirection.stopped {
      setDirection(direction: direction)
    }
  }

  /**
    Sets send state of this `RtpTransceiver`.

    - Parameters:
      - send: Is send direction is enabled.
  */
  func setSend(send: Bool) {
    let direction = self.getDirection()
    var newDirection = RTCRtpTransceiverDirection.stopped
    if send {
      switch direction {
      case .inactive:
        newDirection = RTCRtpTransceiverDirection.sendOnly
      case .sendOnly:
        newDirection = RTCRtpTransceiverDirection.sendOnly
      case .sendRecv:
        newDirection = RTCRtpTransceiverDirection.sendRecv
      case .recvOnly:
        newDirection = RTCRtpTransceiverDirection.sendRecv
      case .stopped:
        newDirection = RTCRtpTransceiverDirection.stopped
      }
    } else {
      switch direction {
      case .inactive:
        newDirection = RTCRtpTransceiverDirection.inactive
      case .sendOnly:
        newDirection = RTCRtpTransceiverDirection.inactive
      case .sendRecv:
        newDirection = RTCRtpTransceiverDirection.recvOnly
      case .recvOnly:
        newDirection = RTCRtpTransceiverDirection.recvOnly
      case .stopped:
        newDirection = RTCRtpTransceiverDirection.stopped
      }
    }

    if newDirection != RTCRtpTransceiverDirection.stopped {
      setDirection(direction: direction)
    }
  }

  /// - Returns: mID of the underlying `RtpTransceiver`.
  func getMid() -> String? {
    if transceiver.mid != nil && transceiver.mid.count == 0 {
      return nil
    }
    return transceiver.mid
  }

  /// - Returns: Preferred `RtpTransceiverDirection` of the underlying `RtpTransceiver`.
  func getDirection() -> TransceiverDirection {
    if self.transceiver.isStopped {
      return TransceiverDirection.stopped
    } else {
      return TransceiverDirection.fromWebRtc(direction: self.transceiver.direction)
    }
  }

  /// Stops the underlying `RTCRtpTransceiver`.
  func stop() {
    self.setDirection(direction: TransceiverDirection.stopped)
  }
}
